[WIP] Support UUID as a choice value for DocumentType

koemeet opened the pull request:
I am trying to make the `DocumentType` support UUIDs too (useful for API's). It is very hacky in my opinion, maybe someone else with more experience could lead me into the right direction.

Related issue #103

koemeet committed:
Make test to check if you can post a UUID as the value of a PHPCR document (phpcr_document form type).

koemeet committed:
Add own IdReader

koemeet committed:
clean up a bit + use idReaders property (small bug).

koemeet committed:
Utitlize UUIDHelper class to determine if we should use the UUID field or not (this also fixes the tests).

steffenbrem commented:
I have a better solution now, by using the `UUIDHelper` class, so based on what kind of value you submit to the form it will check according to that. Only drawback is that I had to copy over the `DoctrineChoiceLoader` and override the `choice_loader` option on the `DocumentType`.

steffenbrem commented:
@dbu Could you give feedback if this is usable?

koemeet committed:
Replace new array syntax with old one to support older PHP versions.

dbu commented on Form/ChoiceList/PhpcrOdmQueryBuilderLoader.php:
```diff
@@ -82,7 +83,7 @@ public function getEntitiesByIds($identifier, array $values)
 
         if (0 == count($values)) {
-            return array();
+            return [];
```
this is only valid with php 5.4+. lets not change that yet, but when we go 2.0 and drop 5.3 support.

dbu commented on Form/Type/DocumentType.php:
```diff
@@ -34,6 +57,68 @@ public function getLoader(ObjectManager $manager, $queryBuilder, $class)
         );
     }
 
+    public function configureOptions(OptionsResolver $resolver)
```
can you add a phpdoc comment explaining why we need to overwrite this?

dbu commented on Form/Type/DocumentType.php:
```diff
@@ -34,6 +57,68 @@ public function getLoader(ObjectManager $manager, $queryBuilder, $class)
+        $choiceLoaders = &$this->choiceLoaders;
+        $type = $this;
+
+        $choiceLoader = function (Options $options) use ($choiceListFactory, &$choiceLoaders, $type) {
```
can you add a comment on the lines where we differ from the orm loader to explain the differences? will make future maintenance easier.

dbu commented on Form/ChoiceList/DoctrineChoiceLoader.php:
```diff
@@ -0,0 +1,194 @@
+ *
+ * @author Steffen Brem <steffenbrem@gmail.com>
+ */
+class DoctrineChoiceLoader implements ChoiceLoaderInterface
```
do we really need to duplicate the whole class? could we extend it and just change what is actually different?

stof commented on composer.json:
```diff
@@ -19,15 +19,13 @@
         "php": ">=5.3.9",
-        "symfony/framework-bundle": "~2.3.27 || ^2.6.6 || ~3.0",
-        "symfony/doctrine-bridge": "~2.3 || ~3.0",
+        "symfony/symfony": "~2.7",
```
this looks wrong to me

steffenbrem commented on Form/ChoiceList/PhpcrOdmQueryBuilderLoader.php:
```diff
@@ -93,8 +94,13 @@ public function getEntitiesByIds($identifier, array $values)
         foreach ($values as $val) {
-            $where->same($val, $alias);
+            if (UUIDHelper::isUUID($val)) {
+                $where->eq()->field($alias.'.uuid')->literal($val)->end();
```
Here I force it to be on the `uuid` property of the Document. Since the `class` passed in the constructor is sometimes `null`, I cant use its metadata. Any ideas on how we could get the propert metadata here?

dbu replied:
we don't have the model instance at this point yet, right? could we find the class from the query builder if its not explicitly specified?

btw, is this assuming that the uuid is mapped on the property `$uuid`? this is an assumption we should not make. maybe we just need an option to specify the uuid property to use with this form type, and if it is not set, we assume path?

steffenbrem replied:
@dbu The class is always in the query builder, but i think we need to use reflection in order to get the class out of it. The issue is more with the tests in `PhpcrOdmQueryBuilderLoaderTest.php`, there the `$class` is never passed, so I could fix this to add a class to the 3rth argument of its constructor.

koemeet committed:
Minimal symfony related bundles and components are 2.7 and also support 3.0; Also worked through code comments and made necessary changes.
Now extends DoctrineChoiceLoader from the doctrine-bridge. Only changed code is now in there.

koemeet committed:
Fixed a bug where if a Document would not have a UUID field mapped, it would still try to use a UUID, but now it will skip it and fall back to using IDs only.

koemeet committed:
Drop Symfony 2.3 in travis CI builds

koemeet committed:
CS fixing

koemeet committed:
CS fixing

koemeet committed:
Add docblock instead of simple comment to explain why we override form options.

koemeet committed:
Use proper class metadata to get the property that is used for the UUID.

koemeet committed:
CS

dbu commented on Form/ChoiceList/PhpcrOdmQueryBuilderLoader.php:
```diff
@@ -93,8 +107,13 @@ public function getEntitiesByIds($identifier, array $values)
+
         foreach ($values as $val) {
-            $where->same($val, $alias);
+            if (UUIDHelper::isUUID($val) && $this->classMetadata) {
```
i would change the order of the checks. and add another check on classMetadata to see if the document is referenceable.

dbu commented on Form/ChoiceList/DoctrineChoiceLoader.php:
```diff
@@ -0,0 +1,126 @@
+            }
+
+            foreach ($values as $i => $id) {
+                if (UUIDHelper::isUUID($id) && null !== $uuidFieldName) {
```
lets switch order here. isUUID is more expensive than checking the field name for null.

dbu commented on Form/ChoiceList/DoctrineChoiceLoader.php:
```diff
@@ -0,0 +1,126 @@
+
+        $uuidFieldName = null;
+        if ($this->classMetadata instanceof PHPCRClassMetadata) {
+            if ($this->classMetadata->referenceable) {
```
lets squash this into one `if` with `&&`.

dbu commented on composer.json:
```diff
@@ -19,15 +19,15 @@
-        "symfony/framework-bundle": "~2.3.27 || ^2.6.6 || ~3.0",
-        "symfony/doctrine-bridge": "~2.3 || ~3.0",
+        "symfony/framework-bundle": "~2.7 || ~3.0",
+        "symfony/doctrine-bridge": "~2.7 || ~3.0",
```
@lsmith77 do we have an issue with this? i guess we are too far with 1.3 to include this PR in this release, so this would affect 1.4 only...

lsmith77 replied:
agreed

Soullivaneuh replied:
Symfony 2.3 is still supported.

AFAIK, @rande decided to keep the support of 2.3 on sonata projects for now.

Why not follow this?

dbu replied:
because it was too much hassle. this will go into the 1.4 version of the
bundle, and we only just released 1.3 today, so we expect symfony 2.3 to
be EOL when this gets released

dbu commented on Tests/Resources/app/config/config.php:
```diff
@@ -5,3 +5,10 @@
+        'csrf_protection' => false,
+    ),
+);
+$container->loadFromExtension('framework', $config);
```
did this and the csrf protection just pop up with the additional tests you do or is this a general fix for a problem?

steffenbrem replied:
I do this since I am submitting a form in the `DocumentTypeTest.php` on line `73`. I don't think we need to have this enabled, so is it okay to disable this globally?

dbu replied:
that is ok then. i was just confused why that was not an issue until
now. but if your test is the first to actually submit a form then it
makes sense. and it does make sense to test actually submitting the form

koemeet committed:
Patched according to comments; Better performance.

steffenbrem commented:
Ready

dbu commented:
yes, please squash the commits into one and then rebase on master (we touched the .travis.yml and composer.json, there will be some conflicts)

dbu commented:
@Soullivaneuh are you familiar with the ORM EntityType? would you maybe have time to look at this pull request and see if you have an idea how to need less code to achieve this? the problem is that with the anonymous callback functions, we need to copy a lot of code to achieve small changes...

Soullivaneuh commented on Form/ChoiceList/DoctrineChoiceLoader.php:
```diff
@@ -0,0 +1,124 @@
+    private $idReader;
+
+    /**
+     * @var null|EntityLoaderInterface
```
`EntityLoaderInterface|null`

Just a detail, but this will keep code consistency. ;-)

Soullivaneuh commented:
@dbu well, I have to admin I don't use DoctrinePHPCRBundle actually.

What do you exactly want from me?

If I'm not wrong, Doctrine ORM support UUID mapping: http://doctrine-orm.readthedocs.org/projects/doctrine-orm/en/latest/reference/basic-mapping.html#identifier-generation-strategies

Would not this be better to propose this PR directly on Symfony Doctrine bridge, especialy for the choice loader?

dbu commented:
thanks for the feedback @Soullivaneuh . 
this bundle builds on top of the bridge. the problem is that this PR is copying a lot of code from there in order to add a few things. but with the callbacks, there seems to be no evident way to avoid copying code. so if you would have an idea how to avoid duplication, that would be awesome.
